profiler/profile.go

```diff
@@ -216,6 +229,108 @@ func goroutineProfile(cfg *config) (*profile, error) {
 	}, nil
 }
 
+func goroutineWaitProfile(cfg *config) (*profile, error) {
+	var (
+		text  = &bytes.Buffer{}
+		pprof = &bytes.Buffer{}
+		start = now()
+	)
+	if err := lookupProfile(GoroutineProfile.String(), text, 2); err != nil {
+		return nil, err
+	} else if err := goroutineDebug2ToPprof(text, pprof); err != nil {
+		return nil, err
+	}
+	end := now()
+	tags := append(cfg.tags, GoroutineWaitProfile.Tag())
+	cfg.statsd.Timing("datadog.profiler.go.collect_time", end.Sub(start), tags, 1)
+
+	return &profile{
+		name: GoroutineWaitProfile.Filename(),
+		data: pprof.Bytes(),
+	}, nil
+}
+
+func goroutineDebug2ToPprof(r io.Reader, w io.Writer) error {
+	goroutines, errs := stackparse.Parse(r)
+
+	functionID := uint64(1)
+	locationID := uint64(1)
+
+	p := &pprofile.Profile{}
+	m := &pprofile.Mapping{ID: 1, HasFunctions: true}
+	p.Mapping = []*pprofile.Mapping{m}
+	p.SampleType = []*pprofile.ValueType{
+		{
+			Type: "waitduration",
+			Unit: "nanoseconds",
+		},
+	}
+
+	for _, g := range goroutines {
+		// TODO(fg) exclude goroutines w/ g.Wait == 0?
+
+		sample := &pprofile.Sample{
+			Value: []int64{g.Wait.Nanoseconds()},
+			Label: map[string][]string{
+				"state": {g.State}, // TODO(fg) split into atomicstatus/waitreason?
+			},
+			NumUnit:  map[string][]string{"goid": {"id"}},
+			NumLabel: map[string][]int64{"goid": {int64(g.ID)}},
+			// TODO(fg) add g.LockedToThread, g.CreatedBy, ...?
+		}
+
+		// Based on internal discussion, the current strategy is to use virtual
+		// frames to indicate truncated stacks, see [1] for how python/jd does it.
+		// [1] https://github.com/DataDog/dd-trace-py/blob/e933d2485b9019a7afad7127f7c0eb541341cdb7/ddtrace/profiling/exporter/pprof.pyx#L117-L121
+		if g.FramesElided {
+			g.Stack = append(g.Stack, &stackparse.Frame{
+				Func: "...additional frames elided...",
+			})
+		}
+
+		for _, call := range g.Stack {
+			function := &pprofile.Function{
+				ID:       functionID,
+				Name:     call.Func,
+				Filename: call.File,
+			}
+			p.Function = append(p.Function, function)
+			functionID++
+
+			location := &pprofile.Location{
+				ID:      locationID,
+				Mapping: m,
+				Line: []pprofile.Line{{
+					Function: function,
+					Line:     int64(call.Line),
+				}},
+			}
+			p.Location = append(p.Location, location)
+			locationID++
+
+			sample.Location = append([]*pprofile.Location{location}, sample.Location...)
+		}
+
+		p.Sample = append(p.Sample, sample)
+	}
+
+	// Put the error message in the pprof profiles as comments in case we need to
+	// debug issues at some point.
+	// TODO(fg) would be nice to also have a metric counter for this
+	if errs != nil {
+		for _, err := range errs.Errors {
+			p.Comments = append(p.Comments, "error: "+err.Error())
+		}
+	}
```

Comment:
@pmbauer we should discuss this!

Reply:
+1 to including the errors in the profile itself
re: metric : do you mean a metric system time series metric, a metric included in the json payload (even system metric) or both?  a statsd counter would be good here (metric system)